File: Miscellaneous/KMP-String-Pattern-Matching/KMPAlgorithm.java
/**
 * KMP (Knuth-Morris-Pratt) String Pattern Matching Algorithm
 * Time Complexity: O(n + m)
 * Space Complexity: O(m)
 */

import java.util.*;

public class KMPAlgorithm {

    // Compute LPS (Longest Prefix Suffix) array
    private static int[] computeLPS(String pattern) {
        int m = pattern.length();
        int[] lps = new int[m];
        int len = 0, i = 1;
        while (i < m) {
            if (pattern.charAt(i) == pattern.charAt(len))
                lps[i++] = ++len;
            else if (len != 0)
                len = lps[len - 1];
            else
                lps[i++] = 0;
        }
        return lps;
    }

    // KMP search algorithm
    public static List<Integer> search(String text, String pattern) {
        List<Integer> res = new ArrayList<>();
        int n = text.length(), m = pattern.length();
        if (m == 0) return res;

        int[] lps = computeLPS(pattern);
        int i = 0, j = 0;

        while (i < n) {
            if (text.charAt(i) == pattern.charAt(j)) {
                i++; j++;
            }
            if (j == m) {
                res.add(i - j);
                j = lps[j - 1];
            } else if (i < n && text.charAt(i) != pattern.charAt(j)) {
                if (j != 0) j = lps[j - 1];
                else i++;
            }
        }
        return res;
    }

    // Interactive mode
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        System.out.print("Enter text: ");
        String text = sc.nextLine();
        System.out.print("Enter pattern: ");
        String pattern = sc.nextLine();

        List<Integer> matches = search(text, pattern);
        if (matches.isEmpty())
            System.out.println("Pattern not found");
        else
            System.out.println("Pattern found at indices: " + matches);
        sc.close();
    }
}

File: Miscellaneous/KMP-String-Pattern-Matching/kmp_algorithm.py
"""
KMP (Knuth-Morris-Pratt) String Pattern Matching Algorithm
Author: Khushi Patel
Time Complexity: O(n + m)
Space Complexity: O(m)
"""

class KMPAlgorithm:
    def compute_lps(self, pattern):
        """Compute Longest Prefix Suffix (LPS) array."""
        m = len(pattern)
        lps = [0] * m
        length, i = 0, 1
        while i < m:
            if pattern[i] == pattern[length]:
                length += 1
                lps[i] = length
                i += 1
            elif length:
                length = lps[length - 1]
            else:
                i += 1
        return lps

    def search(self, text, pattern):
        """Return all starting indices where pattern is found in text."""
        n, m = len(text), len(pattern)
        if m == 0:
            return []

        lps = self.compute_lps(pattern)
        res, i, j = [], 0, 0

        while i < n:
            if text[i] == pattern[j]:
                i += 1
                j += 1
            if j == m:
                res.append(i - j)
                j = lps[j - 1]
            elif i < n and text[i] != pattern[j]:
                j = lps[j - 1] if j else i + 1 - i
                if j == 0:
                    i += 1
        return res


if __name__ == "__main__":
    text = input("Enter text: ").strip()
    pattern = input("Enter pattern: ").strip()

    kmp = KMPAlgorithm()
    matches = kmp.search(text, pattern)

    if matches:
        print("Pattern found at indices:", *matches)
    else:
        print("Pattern not found")
